source/common/upstream/health_discovery_service.cc

```diff
@@ -73,5 +117,86 @@ void HdsDelegate::onRemoteClose(Grpc::Status::GrpcStatus status, const std::stri
   handleFailure();
 }
 
+HdsCluster::HdsCluster(Runtime::Loader& runtime, const envoy::api::v2::Cluster& cluster,
+                       const envoy::api::v2::core::BindConfig& bind_config, Stats::Store& stats,
+                       Ssl::ContextManager& ssl_context_manager,
+                       Secret::SecretManager& secret_manager, bool added_via_api)
+    : info_(new ClusterInfoImpl(cluster, bind_config, runtime, stats, ssl_context_manager,
+                                secret_manager, added_via_api)),
+      runtime_(runtime), initial_hosts_(new HostVector())
+
+{
+  priority_set_.getOrCreateHostSet(0);
+
+  for (const auto& host : cluster.hosts()) {
+    initial_hosts_->emplace_back(HostSharedPtr{new HostImpl(
+        info_, "", resolveProtoAddress2(host), envoy::api::v2::core::Metadata::default_instance(),
+        1, envoy::api::v2::core::Locality().default_instance(),
+        envoy::api::v2::endpoint::Endpoint::HealthCheckConfig().default_instance())});
+  }
+}
+
+ClusterSharedPtr HdsCluster::create() { NOT_IMPLEMENTED; }
+
+void HdsCluster::setHealthChecker(const HealthCheckerSharedPtr& health_checker) {
+  ASSERT(!health_checker_);
+  health_checker_ = health_checker;
+  health_checker_->start();
+}
+
+HostVectorConstSharedPtr HdsCluster::createHealthyHostList(const HostVector& hosts) {
+  HostVectorSharedPtr healthy_list(new HostVector());
+  for (const auto& host : hosts) {
+    if (host->healthy()) {
+      healthy_list->emplace_back(host);
+    }
+  }
+  return healthy_list;
+}
+
+HostsPerLocalityConstSharedPtr HdsCluster::createHealthyHostLists(const HostsPerLocality& hosts) {
+  return hosts.filter([](const Host& host) { return host.healthy(); });
+}
+
+void HdsCluster::initialize(std::function<void()> callback) {
+  ASSERT(initialization_complete_callback_ == nullptr);
+  initialization_complete_callback_ = callback;
+  if (health_checker_) {
+    for (const auto& host : *initial_hosts_) {
+      host->healthFlagSet(Host::HealthFlag::FAILED_ACTIVE_HC);
+    }
+  }
+
+  auto& first_host_set = priority_set_.getOrCreateHostSet(0);
+  auto healthy = createHealthyHostList(*initial_hosts_);
+  ENVOY_LOG(debug, "About to update the priority set.");
+  first_host_set.updateHosts(initial_hosts_, healthy, HostsPerLocalityImpl::empty(),
+                             HostsPerLocalityImpl::empty(), {}, *initial_hosts_, {});
+  initial_hosts_ = nullptr;
+}
+
+const Network::Address::InstanceConstSharedPtr
+HdsCluster::resolveProtoAddress2(const envoy::api::v2::core::Address& address) {
+  try {
+    return Network::Address::resolveProtoAddress(address);
```

Comment:
I think this can just be inlined at the call site, no need for the try/catch case given we know HDS will not be any of the cluster types in the catch case.

Reply:
Done.